### ISSUE TITLE
Choice of numeric or code-based references

### ISSUE THREAD

**Issue by opoudjis** [metanorma/metanorma-nist#110]
From template:

>General Guidance for References and Footnotes:
>References: Authors can choose between using numeric (“[1]”) and alphanumeric (“[SP 800-300]”) references but should use one or the other. 
>Numeric references should always start with [1] and proceed sequentially for each new reference. If beginning a sentence with a numeric reference, start the sentence with “Reference [n]…”
>Alphanumeric references can be useful when there are numerous references to other specifications, and having the document number as part of the reference provides useful context to the reader (so they don’t have to constantly navigate to the references section to see what’s being cited).

This would presumably be done as document attribute. At the moment only alphanumeric references exist, and I'd prefer to keep it that way. If we do use the numeric option, that means that references to standards with standard identifiers will be overridden by numerical references -- unlike what happens in ISO, where the standard identifier is always used in both the Normative References and the Bibliography.

**Comment by opoudjis** [comment]
@ronaldtse What was the decision on this? Are we allowing the option of numeric references?

**Comment by ronaldtse** [comment]
Initially the decision is supposed to be an option -- but it wasn't discussed too deeply. Let me try again.

**Comment by opoudjis** [comment]
#130 implements crude version: if numeric crossreference present on preformatted reference, realise as prefixed number in biblio list, else realise as suffixed docidentifier. If reference is not preformatted, ignore any numeric crossreferencing. This is crude deliberately, since we don't have more explicit guidance from NIST yet.

**Comment by ronaldtse** [comment]
Confirmed, the behavior should be chosen by the author. The reason is backwards compatibility and also a yet to be determined choice for future documents.

**Comment by opoudjis** [comment]
Until now, we allow a special document identifier of  type "metanorma", corresponding to numerical identifiers in a citation. Those document identifiers are treated as bracketed references, and skipped in contexts where a true identifier is called for.

We will extend that class of identifiers to include identifies in references in parentheses. (Cannot use brackets, because of constraints of Asciidoctor syntax). So the following two citations:

````
* [[[ref1,1]]], _citation_
* [[[ref2,(Citn)]]], _citation2_
````

will be rendered with [1] and [Citn], respectively.

**Comment by ronaldtse** [comment]
This is close to being finalized, there are 2 ways of citations (and references listing) in NIST documents (looking ahead):

* Option 1: Numbered reference tag
* Option 2: Named reference tag

NOTE: Unassociated reference lists are disallowed.

## Option 1: Numbered reference tag

The default style.

The numbered reference tag assigned to each reference is in accordance with the order of citation within the document.

Multiple citations to the same reference should be represented re-using the given numeric reference tag. (i.e. if it was "10" in the first citation, it should also be "10" in the subsequent citations)

In the reference list, is sorted by numeric order of the numeric reference tags.

Style: `[#]`

NOTE: The “variant (superscript) style” identified in NIST SP 1041r1 should not be allowed.


## Option 2: Named reference tag

Optional style.

Each reference item is assigned a chosen string reference tag by the author.

In the reference listing, references are ordered by the string reference tags.

NOTE: The sorting algorithm for the reference tags is not 100% settled, it may not be a simple string sort, but we can implement it as such:

Normal sort:
```
[SP800-160]
[SP800-73-4]
```

PubID schema sort:
```
[SP800-73-4]
[SP800-160]
```


## Option 3: Named reference tag, split sections

Optional style. No hanging references are allowed in this style (i.e. all reference items must belong to a reference section).

This is mostly same as Option 2, but the references are allowed to be split into multiple sections. 
This means the Bibliography ("complete reference list") will be composed of multiple sections. 
Authors are able to customize the reference section names.

Within any of the reference sections, the named reference tag ordering is identical to that of Option 2.

There are two acceptable ways to split sections.

### Split method 1: by theme

The order of reference sections is up to the author.

### Split method 2: NIST vs the rest

The order of reference sections is up to the author, but if there is a "NIST-specific" section, this section has to be the first section in the reference list.

All other sections do not need to be ordered.

NOTE 1: Currently, there are some cybersecurity pubs where the NIST pub section is not listed first. However, we should require that if a NIST pub section is included, it must appear as the first section.

NOTE 2: This is commonly used if there are many references but the most important ones are other NIST Special Pubs and FIPS, due to the requirements imposed by them. See SP 800-53, SP 800-171, SP 800-160, and other FISMA-related pubs.

**Comment by opoudjis** [comment]
I will implement the superstructure in isodoc and metanorma-standoc, to insulate it from this complexity. But the complexity seems to be solely a matter of sort orders.

**Comment by opoudjis** [comment]
So, metanorma-standoc and isodoc now cope with both numeric and alphanumeric citations.

All I need to implement is the sort for each section, and it doesn't need to be that big a deal for now:

* If it finds numbers, it sorts them numerically
* If it finds alphanumerics, it sorts them alphanumerically
* If it finds both, well, it will sort them numbers first, but that's ok.

**Comment by opoudjis** [comment]
Realised in rspec, as sort order of bibitems.

**Comment by ronaldtse** [comment]
Thank you @opoudjis , could we also have documentation on how this works?
